### PR TITLE
feat: variable eye to screen distance

### PR DESCRIPTION
**Pull request by jakobchwastek**
## Description

Fixes issue #495 

## Implemented changes

Insert a description of the changes implemented in the pull request.

- Add functionality for distance values in a column to `transforms.pix2deg()`
- Make `distance_cm` optional in `Experiment` and `Screen`
- Add parameter for `distance_column` to  classes `DatasetDefinition`, `GazeDataFrame`.
- Add support for distance column in `GazeDataFrame.transform()`
- Add support for distance column in Dataset loading method `dataset_files.load_gaze_files()` and in loading methods for GazeDataframe `from_numpy()`, `from_pandas()`, `from_csv()`
- Changed a lot of tests..

...

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/pymovements/gaze/experiment.py [hunk 1]
```diff
@@ -56,12 +61,14 @@ def __init__(
             Screen width in centimeters
         screen_height_cm : float
             Screen height in centimeters
-        distance_cm : float
-            Eye-to-screen distance in centimeters
         origin : str
             Specifies the screen location of the origin of the pixel coordinate system.
         sampling_rate : float
             Sampling rate in Hz
+        distance_cm : float | None
```
Comment (dkrako) [inline src/pymovements/gaze/experiment.py:68]: as above

File: src/pymovements/gaze/integration.py [hunk 2]
```diff
@@ -290,7 +309,11 @@ def from_pandas(
         The name of the dva velocity columns in the input data frame.
     acceleration_columns:
         The name of the dva acceleration columns in the input data frame.
-
+    distance_column:
+        The name of the column containing eye-to-screen distance in millimeters for each sample
+        in the input data frame. If specified, the column will be used for pixel to dva
+        transformations. If not specified, the constant eye-to-screen distance will be taken from
+        the experiment definition.
```
Comment (dkrako) [inline src/pymovements/gaze/integration.py:316]: misses a new line before `Returns`

File: tests/gaze/apply_test.py
Comment (dkrako) [inline tests/gaze/apply_test.py:1]: this file should be reverted

File: tests/gaze/detect_test.py
Comment (dkrako) [inline tests/gaze/detect_test.py:1]: this file should be reverted

File: tests/gaze/gaze_dataframe_test.py
Comment (dkrako) [inline tests/gaze/gaze_dataframe_test.py:1]: this file should be reverted

File: tests/plotting/tsplot_test.py
Comment (dkrako) [inline tests/plotting/tsplot_test.py:1]: this file should be reverted

File: tests/plotting/traceplot_test.py
Comment (dkrako) [inline tests/plotting/traceplot_test.py:1]: this file should be reverted

File: tests/plotting/heatmap_test.py
Comment (dkrako) [inline tests/plotting/heatmap_test.py:1]: this file should be reverted

File: tests/gaze/integration_pandas_test.py
Comment (dkrako) [inline tests/gaze/integration_pandas_test.py:1]: this file should be reverted

File: tests/gaze/integration_numpy_test.py [hunk 3]
```diff
@@ -130,13 +135,14 @@ def test_from_numpy_explicit_columns():
         screen_height_px=1024,
         screen_width_cm=38,
         screen_height_cm=30,
-        distance_cm=68,
         origin='lower left',
         sampling_rate=1000.0,
+        distance_cm=68,
```
Comment (dkrako) [inline tests/gaze/integration_numpy_test.py:140]: distance_cm should probably be None then

File: tests/gaze/integration_numpy_test.py [hunk 4]
```diff
@@ -74,23 +75,24 @@ def test_from_numpy_with_schema():
...
 
     experiment = pm.Experiment(
         screen_width_px=1280,
         screen_height_px=1024,
         screen_width_cm=38,
         screen_height_cm=30,
-        distance_cm=68,
         origin='lower left',
         sampling_rate=1000.0,
+        distance_cm=68,
```
Comment (dkrako) [inline tests/gaze/integration_numpy_test.py:87]: distance_cm should probably be None then

File: tests/gaze/gaze_transform_test.py [hunk 5]
```diff
@@ -411,13 +613,13 @@ def fixture_experiment():
...
                 ),
                 'experiment': pm.Experiment(
-                    sampling_rate=1000,
                     screen_width_px=100,
                     screen_height_px=100,
                     screen_width_cm=100,
                     screen_height_cm=100,
-                    distance_cm=100,
                     origin='lower left',
+                    sampling_rate=1000,
```
Comment (dkrako) [inline tests/gaze/gaze_transform_test.py:621]: this change should be reverted

File: tests/gaze/gaze_transform_test.py [hunk 6]
```diff
@@ -373,13 +575,13 @@ def fixture_experiment():
...
```
Comment (dkrako) [inline tests/gaze/gaze_transform_test.py:583]: this change should be reverted

File: tests/gaze/gaze_transform_test.py [hunk 7]
```diff
@@ -336,13 +538,13 @@ def fixture_experiment():
...
```
Comment (dkrako) [inline tests/gaze/gaze_transform_test.py:546]: this change should be reverted

File: tests/gaze/gaze_transform_test.py [hunk 8]
```diff
@@ -298,13 +500,13 @@ def fixture_experiment():
...
```
Comment (dkrako) [inline tests/gaze/gaze_transform_test.py:508]: this change should be reverted

File: tests/gaze/gaze_transform_test.py [hunk 9]
```diff
@@ -257,18 +414,61 @@ def fixture_experiment():
...
+                ),
+                'experiment': pm.Experiment(
                     screen_width_px=100,
                     screen_height_px=100,
                     screen_width_cm=100,
                     screen_height_cm=100,
-                    distance_cm=100,
                     origin='lower left',
+                    sampling_rate=1000,
+                    distance_cm=1,
```
Comment (dkrako) [inline tests/gaze/gaze_transform_test.py:468]: this change should be reverted

**Review by dkrako — CHANGES_REQUESTED** [review]
Great work! Thanks a lot!

I have a major change request, which encompasses several files.
Please revert the changing of the argument order to ensure backwards compatibility. Let's add default arguments instead.

File: src/pymovements/gaze/transforms.py [hunk 10]
```diff
@@ -221,18 +222,41 @@ def pix2deg(
...
+                _arctan2_helper(distance_pixels[component % 2]),
+            ) * (180 / np.pi)
+            for component in range(n_components)
+        ]
+
+    elif isinstance(distance, str):
+        # distance is a column name
+        degree_components = [
+            # True division by 10 is needed to convert distance from mm to cm
+            pl.arctan2(
```
Comment (dkrako) [inline src/pymovements/gaze/transforms.py:246]: can you reuse pl.arctan2() for the scalar case above?

maybe similar to this? 

```python
if isinstance(distance, (float, int)):
    distance_series = pl.lit(distance)
elif isinstance(distance, str):
    # True division by 10 is needed to convert distance from mm to cm
    distance_series = pl.col(distance).truediv(10)
else:
    raise TypeError()

distance_pixels = distance_series.mul(
    screen_resolution[component % 2] / screen_size[component % 2],
)  

degree_components = [
    pl.arctan2(centered_pixels.list.get(component), distance_pixels) * (180 / np.pi)
    for component in range(n_components)
]

You can then completely remove `_arctan2_helper()`

File: tests/gaze/screen_test.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+def test_dva_properties_with_no_distance_cm(property_name):
+    screen = pm.Screen(1920, 1080, 30, 20, None, 'lower left')
+    with pytest.raises(ValueError):
+        getattr(screen, property_name)
+
+
+@pytest.mark.parametrize('property_name', ['x_max_dva', 'y_max_dva', 'x_min_dva', 'y_min_dva'])
+def test_dva_properties_with_distance_cm(property_name):
+    screen = pm.Screen(1920, 1080, 30, 20, 60, 'lower left')
+    try:
```
Comment (dkrako) [inline tests/gaze/screen_test.py:36]: why is the try/except needed?

**Review by dkrako — CHANGES_REQUESTED** [review]
I have one small request for a final refactoring, and I'm confused about the try/except clauses in some tests

File: tests/gaze/screen_test.py (hunk 11 above)
Comment (jakobchwastek) [reply tests/gaze/screen_test.py]: honestly idk. I removed it. Is it sufficient to just test if call is not failing?

**Comment by jakobchwastek** [comment]
Thanks a lot for your comments @dkrako. Please let me know if any further change is required.

File: tests/gaze/screen_test.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+def test_screen_pix2deg_with_no_distance_cm():
+    screen = pm.Screen(1920, 1080, 30, 20, None, 'lower left')
+    with pytest.raises(ValueError):
+        screen.pix2deg([[0, 0]])
+
+
+def test_screen_pix2deg_with_distance_cm():
+    screen = pm.Screen(1920, 1080, 30, 20, 60, 'lower left')
+    try:
```
Comment (dkrako) [inline tests/gaze/screen_test.py:48]: one last try-catch left

**Review by dkrako — CHANGES_REQUESTED** [review]
just one itsy bitsy teenie weenie change request

**Review by dkrako — APPROVED** [review]
Thanks a lot for this great work!

**Pull request merged**